pkg/util/metric/metric_test.go

```diff
@@ -568,3 +603,212 @@ func TestMergeWindowedHistogram(t *testing.T) {
+				}, 10)
+
+				c.Update(map[string]string{
+					"label1": "value3",
+					"label2": "value4",
+				}, 10)
+
+				// overwrite the previous value
+				c.Update(map[string]string{
+					"label1": "value3",
+					"label2": "value4",
+				}, 20)
+			})
+		})
+
+		t.Run("inc", func(t *testing.T) {
+			assert.NotPanics(t, func() {
+				c.Inc(map[string]string{
+					"label1": "value1",
+					"label2": "value2",
+				}, 10)
+
+				c.Inc(map[string]string{
+					"label1": "value3",
+					"label2": "value4",
+				}, 10)
+			})
+		})
+
+		t.Run("count", func(t *testing.T) {
+			assert.Equal(t, int64(20), c.Count(map[string]string{
+				"label1": "value1",
+				"label2": "value2",
+			}))
+
+			assert.Equal(t, int64(30), c.Count(map[string]string{
+				"label1": "value3",
+				"label2": "value4",
+			}))
+
+			// timeseries doesn't exist
+			assert.Equal(t, int64(0), c.Count(map[string]string{
+				"label1": "value5",
+				"label2": "value6",
+			}))
+		})
+
+		t.Run("to prometheus metrics", func(t *testing.T) {
+			c.assertPrometheusMetrics(t, []toPromMetricsTC{{
+				labels: [][]string{{"label1", "value1"}, {"label2", "value2"}},
+				value:  20,
+			}, {
+				labels: [][]string{{"label1", "value3"}, {"label2", "value4"}},
+				value:  30,
+			}})
+		})
+	})
+
+	t.Run("labels provided exceed what is declared", func(t *testing.T) {
+		c := NewCounterVec(emptyMetadata, []string{"label1", "label2"})
+		t.Run("update", func(t *testing.T) {
+			assert.NotPanics(t, func() {
+				c.Update(map[string]string{
+					"label1": "value1",
+					"label2": "value2",
+					"label3": "value3",
+				}, 10)
+
+				c.Update(map[string]string{
+					"label1": "value1",
+					"label2": "value2",
+					"label3": "value3",
+					"label4": "value4",
+					"label5": "value5",
+				}, 50)
+			})
+		})
+
+		t.Run("count", func(t *testing.T) {
+			assert.NotPanics(t, func() {
+				assert.Equal(t, int64(50), c.Count(map[string]string{
+					"label1": "value1",
+					"label2": "value2",
+					"label3": "value3",
+					"label6": "value6",
+				}))
+			})
+		})
+
+		t.Run("to prometheus metrics", func(t *testing.T) {
+			c.assertPrometheusMetrics(t, []toPromMetricsTC{
+				{
+					labels: [][]string{{"label1", "value1"}, {"label2", "value2"}},
+					value:  50,
+				},
+			})
+		})
+		// we don't have to test all operation again
+	})
+
+	t.Run("labels provided are less than what is declared", func(t *testing.T) {
+		c := NewCounterVec(emptyMetadata, []string{"label1", "label2", "label3"})
+		t.Run("update", func(t *testing.T) {
+			assert.NotPanics(t, func() {
+				c.Update(map[string]string{
+					"label1": "value1",
+					"label2": "value2",
+				}, 10)
+			})
+		})
+
+		t.Run("count", func(t *testing.T) {
+			assert.Equal(t, int64(10), c.Count(map[string]string{
+				"label1": "value1",
+				"label2": "value2",
+				"label3": "",
+			}))
+
+			assert.Equal(t, int64(0), c.Count(map[string]string{
+				"label1": "value1",
+				"label2": "value2",
+				"label3": "value3",
+			}))
```

Comment:
there should be a test case to increment just label1, and we should assert that a separate set of time series metrics get persisted.